File: all/1128_number_of_equivalent_domino_pairs/sol_better.c
static inline int max(const int i1, const int i2) {
    return ((i1 > i2) ? i1 : i2);
}

static inline int min(const int i1, const int i2) {
    return ((i1 < i2) ? i1 : i2);
}

int numEquivDominoPairs(int** dominoes, int dominoesSize, int* dominoesColSize) {
    int domino_values_counter[100] = {0};
    int domino_value = 0;
    int result_pairs_count = 0;

    /* Compute each domino's value and their count. */
    for (int i = 0; i < dominoesSize; ++i) {
        /*
         * Use normalization to make a domino like (a, b) and the rotated
         * domino like (b, a) can have the same domino value.
         */
        domino_value = (
            max(dominoes[i][0], dominoes[i][1]) * 10
            +
            min(dominoes[i][0], dominoes[i][1])
        );

        ++domino_values_counter[domino_value];
    }

    /*
     * After computing each domino value's count, we can use that information
     * to compute how many domino pairs are the same.
     */
    for (int i = 0; i < 100; ++i) {
        /*
         * If a domino value is x, there can be
         * ((1 + (x - 1)) * (x - 1)) / 2 pairs of dominoes are considered
         * equivalent.
         */
        result_pairs_count += (
            (
                (1 + (domino_values_counter[i] - 1))
                *
                (domino_values_counter[i] - 1)
            )
            >> 1
        );
    }

    return result_pairs_count;
}

File: all/1128_number_of_equivalent_domino_pairs/sol_better.py
from typing import List


class Solution:
    def numEquivDominoPairs(self, dominoes: List[List[int]]) -> int:
        domino_values_counter_list = [0] * 100

        # Compute each domino's value and their count.
        for a, b in dominoes:
            # Use normalization to make a domino like (a, b) and the rotated
            # domino like (b, a) can have the same domino value.
            domino_values_counter_list[(max(a, b) * 10) + min(a, b)] += 1

        # After computing each domino value's count, we can use that information
        # to compute how many domino pairs are the same.
        # If a domino value's count is x, there can be
        # ((1 + (x - 1)) * (x - 1)) / 2 pairs of dominoes are considered
        # equivalent.
        return sum(
            (((1 + (count - 1)) * (count - 1)) >> 1)
            for count in domino_values_counter_list
        )


if '__main__' == __name__:
    sol = Solution()
    testcases_list = [
        [[[1, 2], [2, 1], [3, 4], [5, 6]]],
        [[[1, 2], [1, 2], [1, 1], [1, 2], [2, 2]]],
    ]

    for testcase in testcases_list:
        print(sol.numEquivDominoPairs(*testcase))
